### DISCUSSION TITLE
DockerOperator files across containers

### DISCUSSION THREAD

**Discussion by AbdAlRahman-Odeh-99** [apache/airflow#26064] [Q&A]
Hello,
Let's say I have a task that uses dockerOperator to prepare create directories. (Ex: /mnt/vol/data)
After that, I have another task that uses dockerOperator to perform certain functionalities on the data that was created by the previous dockerOperator. (Ex: Needs access to /mnt/vol/data that was created)
Of course, each container has it's own files and domain, thus (/mnt/vol/data) can't be reached by the second container. Is there a way to make this possible?
Furthermore, is there a way I could copy files from the container to my host?

**Comment by potiuk** [comment]
Plenty of options:

1) you can bind mount volumes if your docker engine runs on the same machine (but that requires running under local executor - otherwise your machines might be different for each task

2) You can have a shared filesystem (NFS or the like) that you can mount but for that your docker engine needs to support it (This is for example what PVC in Kubernetes does). Consult your container docs for that (and pass the right `mounts` parameter. This is ost transparent if you want to use DockerOperator - if you use same mounts for different DockerOperators, they will share the data automatically.

3) If you have remote engine, you can send files to the docker engine before/after with the `docker cp` command https://docs.docker.com/engine/reference/commandline/cp/  but again this will only work in local executor if you want to use Docker Operator because each task can be executed on different machine if non-LocalExecutor is used.

4) you can use @task decorator and utilise DockerHook to run your docker commands instead - then in the @task decorated method you could do several commands - including `docker cp` back and forth before/after running each command. Docker Hook get_conn() method returns the usual Python Docker Client that you can do whatever with - no limits, whatever Docker Client allows you to do you can do in Python and you can run multiple commands. This is useful if you have some auth params/secrets for authentication that you would liek to utilise when instantiating the hook (when creating the DockerHook you can pass docker_conn_id for example).  

5) You can write your docker push/execute/pull dance inside plain Python Operator or BashOperator as sequence of commands to execute if  you do not need Hook authentication.

**Reply by AbdAlRahman-Odeh-99** [reply]
Dear Potiuk,
Thank you for your answer
Currently, I am running Airflow with SequentialExecuter, sqlite database and executing it on my local machine.

For example, I have the following DAG:
```
    t1 =  DockerOperator(
        task_id='t1',
        image='ubuntu:latest',
        api_version='auto',
        auto_remove=True,
        docker_url='unix://var/run/docker.sock',
        network_mode='bridge',
        command="mkdir -p /mnt/vol",  
    )

    t2 = DockerOperator(
        task_id = "t2",
        image='ubuntu:latest',
        api_version='auto',
        auto_remove=True,
        docker_url='unix://var/run/docker.sock',
        network_mode='bridge',
        command='cd /mnt/vol'
 )
```
Of course, this will result with an error of ```cd: no such file or directory```. So what is the simplest way to let ```t2``` have access to ```t1``` without the need to change the executor?

**Reply by potiuk** [reply]
First of all using sequential executor is only for development - if you need to  move it production later - don't use it. It's only for quick testing. 

Regarding the question - you need ot explore the options that I outlined above.

**Reply by potiuk** [reply]
``mounts`` parameter is your friend
